---
title: "An open and continuously updated fern tree of life (FTOL)"
mainfont: Roboto-Regular.ttf
fontsize: 11pt
mainfontoptions: 
  - BoldFont=Roboto-Bold.ttf
  - ItalicFont=Roboto-Italic.ttf
  - BoldItalicFont=Roboto-BoldItalic.ttf
output: 
  bookdown::pdf_document2:
    citation_package: default
    toc: no
    number_sections: yes
    fig_caption: yes
    keep_tex: no
    latex_engine: xelatex
    pandoc_args: [ "--csl", "frontiers.csl"]
  bookdown::word_document2:
    pandoc_args: [ "--csl", "frontiers.csl"]
    toc: no
    number_sections: no
    reference_docx: Frontiers_Template_SI.docx
header-includes: 
  - \usepackage{float} # keep figures from floating away
  - \makeatletter\renewcommand*{\fps@figure}{H}\makeatother # keep figures from floating away
  - \usepackage{fancyhdr} # enable header
  - \cfoot{\thepage} # footer
editor_options: 
  chunk_output_type: console
params:
  doc_type: manual
---

---
bibliography: 
- `r here::here("ms/references.yaml")`
- `r here::here("ms/references_other.yaml")`
- `r here::here("ms/references_other.bib")`
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
	echo = FALSE, warning = FALSE, message = FALSE, results = 'hide')
# Load packages
source(here::here("R/ms_packages.R"))
```

```{r formatters-load}
# Load formatting functions defined in functions.R
knitr::read_chunk(
	here::here("R/functions.R"),
	labels = "formatters",
	from = "formatters start",
	to = "formatters end"
)
```

```{r, formatters}
```

```{r load-functions, eval = params$doc_type == "manual"}
# Load functions (only if running interactively)
source(here::here("R/functions.R"))
```

```{r load-targets}
# Targets from FTOL workflow

# Specify path to FTOL cache
# FIXME: change to ftol_cache when ready
ftol_cache <- here::here("working/_targets")

tar_load(
	c(
		accs_exclude,
		fern_sanger_seqs_raw,
		raw_fasta_all,
		sanger_seqs_combined_filtered,
		sanger_seqs_rogues_removed,
		sanger_accessions_selection,
		plastome_seqs_combined_filtered,
		plastome_metadata_raw_all,
		plastome_metadata_renamed,
		sanger_alignment,
		plastome_alignment,
		plastome_metadata_renamed,
		ppgi_taxonomy,
		plastid_calibration_dates,
		pteridocat,
		sanger_alignment_tbl,
		plastome_alignment_tbl,
		monophy_by_clade,
		plastome_tree,
		plastid_tree_dated,
		fossil_calibration_tips,
		ts_fossil_calibration_tips,
		treepl_cv_results),
	store = ftol_cache
)

# Targets from MS workflow
tar_load(
	c(
		accepted_species,
		sanger_sampling,
		coverage_by_rank,
		total_coverage,
		bb_coverage_by_rank,
		bb_total_coverage,
		gb_species_by_year,
		ref_files,
		other_dates,
		crown_ages,
		fossil_ferns_raw,
		family_stem_ages,
		fern_monophy_summ_tbl
	),
	store = here::here("_targets")
)
```

```{r abstract-stats}
# Make tibble of outgroup species
og_species <- 
	plastome_metadata_renamed %>%
	select(species, outgroup) %>%
	filter(outgroup == TRUE)

# Count number of fern species in Sanger data
n_sanger_ferns <-
	sanger_sampling %>% 
	filter(outgroup == FALSE) %>%
	n_distinct(.$species)

# Count number of calibration points
n_calibrations_ferns <-
	fossil_calibration_tips %>%
	assert(is_uniq, node_calibrated) %>%
	filter(!affinities %in% c("Tracheophytes", "Euphyllophytes")) %>%
	pull(node_calibrated) %>%
	n_distinct() %>%
	number()

n_calibrations_total <-
	fossil_calibration_tips %>%
	assert(is_uniq, node_calibrated) %>%
	pull(node_calibrated) %>%
	n_distinct() %>%
	number()
```

<!-- Set left justification -->
\raggedright

<!-- Don't indent -->
\setlength{\parindent}{0ex}

**Joel H. Nitta^1^, Eric Schuettpelz^2^, Santiago Ramírez-Barahona^3^, and Wataru Iwasaki^1,4,5,6,7,8^**

^1^Department of Biological Sciences, Graduate School of Science, The University of Tokyo, Tokyo, Japan

^2^Department of Botany, National Museum of Natural History, Smithsonian Institution, Washington, D.C., USA

^3^Departamento de Botánica, Instituto de Biología, Universidad Nacional Autónoma de México, Mexico City, Mexico

^4^Department of Integrated Biosciences, Graduate School of Frontier Sciences, The University of Tokyo, Chiba, Japan

^5^Department of Computational Biology and Medical Sciences, Graduate School of Frontier Sciences, The University of Tokyo, Chiba, Japan

^6^Atmosphere and Ocean Research Institute, The University of Tokyo, Chiba, Japan

^7^Institute for Quantitative Biosciences, The University of Tokyo, Tokyo, Japan

^8^Collaborative Research Institute for Innovative Microbiology, The University of Tokyo, Tokyo, Japan

**\*Correspondence:**    
Joel H. Nitta  
joelnitta@gmail.com

Keywords: Fern~1~, Phylogeney~2~, Plastome~3~, PPGI~4~, Pteridophyte~5~, *rbcL*~6~.

# Abstract

Ferns (ca. 12,000 spp.) are the second most diverse clade of vascular plants after seed plants and have received intense focus from molecular systematists, resulting in the publication of molecular trees for nearly all major clades and DNA sequences for more than one third of all fern species deposited in GenBank.
Global trees including all ferns have been published periodically, but as molecular systematics research continues at a rapid pace, such trees quickly become outdated.

Here, we develop a mostly automated, reproducible, open pipeline to generate a continuously updated fern tree of life (FTOL) from DNA sequence data on GenBank.
Our tailored sampling strategy combines whole plastomes (few taxa, many loci) with Sanger-sequenced loci (many taxa, few loci) to obtain a global fern phylogeny with high resolution along the backbone and maximum sampling across the tips.
We use an expert-curated reference taxonomy to resolve synonyms in compliance with Pteridophyte Phylogeny Group I (PPGI), the community-driven taxonomic system widely accepted by many pteridologists.

The current FTOL includes `r comma(n_sanger_ferns)` species, an increase of more than 20% relative to the previous largest fern phylogeny.
We also used an updated and expanded list of `r n_calibrations_ferns` fern fossil calibration points to date the tree, and estimated ages for most families and deeper clades to be significantly older compared to previous studies.

FTOL and its accompanying datasets including the fossil list and taxonomic database will be updated on a semi-annual basis and be made available via a web portal and R packages, enabling researchers to have immediate access to the most recent, comprehensively sampled fern phylogeny.
FTOL will be useful for anybody studying this important group of plants over a wide range of taxonomic scales from smaller clades to the entire tree.
We anticipate FTOL will be particularly relevant for macroecological studies at the regional to global scale and will inform future taxonomic systems with the most recent hypothesis of fern phylogeny.

# Introduction

Ferns (ca. 12,000 species) are the second most diverse clade of vascular plants after seed plants (ca. 300,000 species), and are a useful study system for understanding processes of biogeography [@Tryon1986; @Kato1993], community ecology [e.g., @Hennequin2014; @Lehtonen2015], and speciation [e.g., @Kao2020].
A well-sampled molecular phylogeny is key to any investigation of fern evolutionary biology.
Fortunately, ferns have received relatively intense focus from molecular systematists, resulting in the publication of molecular trees for nearly all major clades and DNA sequences for more than one third all currently recognized fern species deposited in GenBank.
Thus, there is both sufficient sampling and a pressing need for a globally sampled fern phylogeny.

Past efforts to construct such a global phylogeny have steadily increased in number of species as sequences in GenBank have accumulated [@Pryer2004a; @Lehtonen2011; @Schuettpelz2007; @Testo2016a].
Indeed, accumulation of plastid fern accessions in GenBank show no sign of slowing since the most recent global phylogeny of ferns [@Testo2016a\; `r figure("genbank")`].
There is a need therefore, not only for a global fern phylogeny, but one that is self-updating to keep pace with the continued accumulation of molecular data.
This will eliminate the need for researchers to "rebuild the wheel" each time they want to use a globally sampled fern phylogeny.

Multiple efforts exist to automatically or semi-automatically generate trees for any arbitrary part of the tree of life [@Antonelli2016], a tree for all plants [@Eiserhardt2018], or even the entire tree of life at once [@Hinchliff2015] which would subsume a global fern phylogeny.
While these may be appropriate in some cases, they may not produce an optimal fern phylogeny due to the need to take "one-size-fits-all" approaches to accommodate such a wide phylogenetic breadth. 
By focusing methods and datasets specifically on ferns, it should be possible to generate a higher quality end-product (tree) that can be used "as-is" by biologists studying these organisms.
Furthermore, there is much to be gained from integrating a carefully designed global fern phylogeny with the fern systematics community that would not be possible with a "tree of all life" or "tree of all plants".

Recently, a genus-level taxonomy of ferns and lycophytes was established following a open, community-driven model [@PteridophytePhylogenyGroupI2016\; hereafter "PPGI"].
Thanks to this community-first approach, PPGI has been widely accepted and used (cited 881 times on Google Scholar as of writing).
However, there were clearly problematic (non-monophyletic) genera at the time PPGI was published, and many new taxonomic changes have been (and will continue to be) made since [e.g., @Almeida2017; @Shang2018; @Zhang2020].
It is anticipated that "PPGII" will be an online, open resource that can be updated as necessary (Schuettpelz, pers. comm.).
We envision an open, continuously updated global fern phylogeny that is directly integrated with PPGII so that taxonomic decisions reflecting phylogeny can be made based on community consensus using the most recently available data.

Here, we leverage our specialized taxonomic knowledge to design a custom, fully reproducible, mostly automated pipeline to generate a maximally sampled global fern tree of life (FTOL).
We plan to run the pipeline on a semi-annual basis and make the results freely available online through a web portal and R package.
This will enable anybody interested in the biology of ferns to have access to the most recent hypothesis of fern phylogeny.
We anticipate FTOL will have several impacts for the field of fern systematics and evolution: 1) it will always provide the most up-to-date snapshot of our collective understanding of fern phylogeny; 2) it will allow for continuous assessment of taxonomy, and indicate those parts of the tree that are in need of taxonomic revision; 3) it will be an important source of data for phylogenetic comparative and macroevolutionary studies of ferns.

# Materials and Methods

## Locus selection

The plastid genome has been the most widely sequenced genomic compartment in ferns by far (`r figure("genbank")`), so we used plastid loci to build our tree.
Our sampling includes two major categories of sequence data: 1) seven loci (*atpA*, *atpB*, *matK*, *rbcL*, *rps4*, *trnL*--*trnF*, and *rps4*--*trnS*) that have been frequently used for molecular systematics of ferns and are typically obtained by PCR and Sanger sequencing ("Sanger loci"); and 2) a much larger set of single-copy, coding genes typically obtained from next-generation sequencing of the plastome ("plastome loci").
Here, the locus *trnL*--*trnF* includes the *trnL* intron, the 3' *trnL* exon, and the spacer between *trnL* and *trnF*; the locus *rps4*--*trnS* includes the spacer between *rps4* and *trnS*.
The set of plastome loci is based on the list of 83 protein-coding genes of @Wei2017a, which was filtered to only genes that show no evidence of duplication (77 genes), then combined with the two spacers (79 loci total).
The plastome loci include all seven Sanger loci.

## Dataset construction

All sequences were downloaded from GenBank.
GenBank queries were formatted including locus name, "Polypodiopsida[ORGN]", and "[PDAT]" to specify sequence date range (e.g., 1980/01/01[PDAT]:2021/12/12[PDAT]) so that a given query always results in the same set of sequences regardless of when the query is made.
Names including the terms "aff." or "cf.", hybrid formulas, and environmental DNA samples were excluded.
There is no formal definition of genomic vs. Sanger sequences in GenBank, so we used an empirical sequence length cutoff to distinguish between Sanger (=< `r number(7000)` bp) or plastome sequences (> `r number(7000)` bp) with the "[SLEN]" term.

There is a lack of consensus on sequence annotation in GenBank; the same locus may be annotated using different names, or not annotated at all.
To avoid missing sequences due to differences in annotation format, we used the "Reference_Blast_Extract.py" script of  superCRUNCH [@Portik2020] to extract target loci from GenBank FASTA files.
Briefly, this involves querying candidate FASTA files downloaded from GenBank with BLAST [@Altschul1997] against a reference database of full length, representative sequences [i.e., a "baited search"\; @Smith2009b; @Smith2019].
Those portions of the query that have a significant match in the reference are extracted and written to a new filtered FASTA file.

We constructed the superCRUNCH reference databases by first downloading fern sequences from GenBank, then extracting target gene sequences with a custom R script that parses the GenBank flatfile; this only works for properly annotated accessions.
We then filtered the sequences to a single representative longest sequence per genus.
Next, we aligned the filtered sequences with MAFFT [@Katoh2002] and removed poorly aligned regions with trimAl [@Capella-Gutierrez2009].
To maximize the size of the reference database for Sanger loci, we then ran "Reference_Blast_Extract.py" using these sequences as references, followed by filtering to the longest sequence per genus and alignment and cleaning as before; this retrieved additional sequences that lacked annotations in the first round.
The cleaned alignments were then used as references for superCRUNCH to obtain a maximally sampled set of fern sequences from sequences downloaded from GenBank.

## Taxonomic name resolution

One goal of this project is to generate a phylogeny that is consistent with PPGI.
Species names in GenBank, which use the NCBI taxonomy [@Federhen2012; @Schoch2020], do not necessarily conform to PPGI.
Furthermore, the NCBI taxonomy is not curated specifically for ferns and includes many fern synonyms.
Therefore, we standardized all species names in the GenBank sequences against a reference taxonomy.
We designed the reference taxonomy for this project based on the World Ferns database v.12.8 [@Hassler2022], which conforms to PPGI (with some exceptions explained below) and is available in Darwin Core format [@DarwinCore2009] via the Catalog of Life [@COL2366].

We used the taxastand R package [@Nitta2021] to resolve species names, which can account for differences in formatting of taxonomic authors (`r eg`, parenthetical basionym author present or absent) as well as perform fuzzy matching, which is needed to account for spelling errors or variations in author names.
We manually inspected any fuzzily matched or non-matching names.
This revealed some species names in GenBank that were lacking in the World Ferns database, spelling errors, as well as some names in the database that needed to be treated differently (`r eg`, changes in synonymy).
Therefore, we updated and edited the World Ferns database using the dwctaxon R package (FIXME add citation), which is designed to work with taxonomic data in the Darwin Core format.
We refer to the resulting taxonomic database as "pteridocat", and have made it freely available online (https://github.com/joelnitta/pteridocat) so that other researchers may standardize taxonomic names in their data to match those of FTOL.

There are differences between pteridocat and PPGI, mostly at the genus level.
In the time since PPGI was published, several new genera have been published.
Furthermore, multiple genera included in PPGI were known to be non-monophyletic and included on a tentative basis [@PteridophytePhylogenyGroupI2016].
Pteridocat includes newly published genera and some genera that were not recognized by PPGI as well as nothogenera, which were not included in PPGI [@Liu2020a].
Differences between pteridocat and PPGI are summarized in `r s_table("genera_diff")`.

## Removal of rogue sequences using BLAST

Another potential issue with GenBank sequences is the presence of misidentified sequences, poor quality sequences, and contaminations (hereafter collectively "rogues").
We removed putative rogues from Sanger sequences as follows.
First, we constructed a BLAST library including all extracted Sanger sequences.
Next, we conducted one BLAST search for each Sanger sequence against the library (all-by-all BLAST).
We compared the families (following PPGI) of the matching sequences to each query; in the case the top three best matches belonged to a different family than the query, that query was considered a contamination and excluded from further analysis.
For Cyatheales and Saccolomatineae, which include several closely related small families, we used the order and sub-order level, respectively, instead of family.
Species belonging to monotypic families were not considered for this filtering.
While this method cannot account for rogues at finer taxonomic levels, it is an efficient method to remove obviously incorrect sequences.
We inspected all accessions flagged this way as rogues before excluding them.
In the case that that mismatch of family was due to incorrect taxonomy (`r eg`, species name on GenBank matches the correct family but name used in the taxonomic database does not), we updated the taxonomic database accordingly.

## Sequence concatenation and selection

A typical step in multilocus workflows is to concatenate loci across samples.
For phylogenetic studies aiming to produce one tip per species, this is often done by concatenating the longest sequence per locus within each species.
However, such an approach is potentially problematic because accessions on GenBank may be misidentified and/or species may not be monophyletic.
Therefore, we concatenated accessions and selected one final set of concatenated loci per species as follows (here, we refer to "accession" to mean a single locus within a GenBank accession; GenBank accessions may have multiple loci per accession, but we already split those out using superCRUNCH as described previously).
First, we constructed gene trees with FastTree [@Price2009; @Price2010] on default settings, and classified species as monophyletic if they were monophyletic in all gene trees including multiple accessions of that species.
We then concatenated loci that met any of the following three conditions: 1) if a species was monophyletic, loci were concatenated by selecting the longest accession per locus within that species; 2) if all accessions originated from the same voucher specimen, loci were concatenated across accessions; 3) if all accessions for a given species originated from only one publication, loci were concatenated across accessions.
Accessions not meeting any of these conditions were not concatenated.
All calculations of sequence length excluded missing bases.

We then selected the final set of concatenated loci for each species based on presence of *rbcL* and sequence length.
We sought to maximize representation of *rbcL* as this is historically the most sequenced locus for ferns, and maximal sampling of one locus has been shown to improve results in super-matrix phylogenies [@Talavera2021].
Our procedure for selecting accessions for each species is as follows.

1. If accessions are concatenated including *rbcL* and at least one other locus, select the set of concatenated accessions with the greatest total sequence length.
2. Otherwise, if accessions include *rbcL*, select the accession with the longest *rbcL* sequence.
3. Otherwise, select the accession or set of concatenated accessions with the greatest total sequence length.

The above steps were not needed for plastome data, as each plastome sequence on GenBank originates from a single voucher specimen.
For these, we selected one specimen per species with the longest combined sequence length across all plastome loci.

## Sequence alignment

For non-spacer regions, we aligned each locus separately with MAFFT with automatic adjustment for sequence direction and other settings on default. Spacer regions are difficult to align across higher taxonomic levels within ferns (e.g., family or higher) as they include frequent indels; however, spacer regions are very useful for phylogenetic analysis at finer levels (`r eg`, within genera or family) where slower-evolving genes like *rbcL* may not provide enough resolution.
Therefore, we first aligned spacer regions for each family with MAFFT, then merged the subalignments using the MAFFT "\-\-merge" option.
This retains the indels within each subalignment while aligning across subalignments.
For some small families with fewer than three species each that could be used for subalignments, these were added as "singletons" during the MAFFT "\-\-merge".
Anemiaceae showed an extremely high number of indels compared to other families and could not be reliably aligned across families, so we excluded it from the spacer region data.
We also excluded outgroups from the spacer region data as they cannot be reliably aligned to ferns.

Next, we removed poorly aligned regions from the alignments using trimAl.
We removed regions with >1% or >5% of sequences having gaps from coding regions or spacers, respectively.

## Phylogenetic analysis

### Backbone phylogeny

We generated a backbone phylogeny using maximum likelihood (ML) analysis of the concatenated plastome dataset in IQ-TREE [@Nguyen2015].
ModelFinder [@Kalyaanamoorthy2017] was implemented in IQ-TREE to select the best-fitting model of sequence evolution.
To reduce computational burden, we only tested models based on the General Time Reversible (GTR) model [@Tavare1986] and did not partition the dataset.
The best-fitting model was selected automatically by IQ-TREE according to Bayesian Information Criterion (BIC) score.
Node support was assessed with `r number(1000)` ultrafast rapid bootstrap replicates [@Minh2013]. 

### Initial Sanger phylogeny

Next, we used the backbone phylogeny as a constraint tree to conduct initial phylogenetic analysis of the concatenated Sanger dataset in IQ-TREE with the "-fast" option and the GTR+I+G model.
We prioritized computation speed for this step, as we needed to repeat it multiple times as described in the next section.

### Removal of rogue sequences based on initial Sanger phylogeny

We inspected the initial Sanger phylogeny to identify any remaining rogues or names that needed updating in the taxonomic database.
We checked for monophyly at taxonomic levels at or above genus using the R package "MonoPhy" [@Schwery2016].
We updated our taxonomic database if the molecular data clearly indicated the current usage of a synonym was incorrect [`r eg`, a taxonomic intruder into an otherwise expected monophyletic genus with a synonym available for that genus\; expected monophyly follows @PteridophytePhylogenyGroupI2016].
However, this was not possible in all cases, particularly in groups that are in need of taxonomic revision and are known to include non-monophyletic genera (`r eg`, cheilanthoid ferns, grammitid ferns, microsoroid ferns).
While we consider our phylogeny may serve as a guide for future taxonomic revisions, we did not make any taxonomic changes that would require valid publication according to the International Code of Nomenclature for Algae, Fungi, and Plants.

Based on the results of this inspection, we updated the list of rogue accessions to exclude, updated the taxonomic database, then re-ran all analyses up to this step.
When re-running the analysis, it is possible that there were newly added GenBank accessions in place of the excluded rogues that themselves are rogues.
Therefore, we repeated this process until the monophyly of all higher-level (`r eg` genus rank and higher) taxa that were expected to be monophyletic according to @PteridophytePhylogenyGroupI2016 was either confirmed or could not be achieved due to outstanding taxonomic issues.

### Final Sanger phylogeny

We generated the final Sanger phylogeny by using the backbone phylogeny as a constraint tree in ML analysis of the final concatenated Sanger dataset with the same model selection procedure and bootstrapping as used for the backbone phylogeny. We did not conduct a more exhaustive analysis testing various partitions of the data or other phylogenetic inference methods or models as our goal is to produce a single species-level phylogeny that is a reasonable hypothesis of fern evolution, not to interrogate the outcomes of multiple, more or less equally applicable methods.

## Molecular dating

```{r dating-methods-stats}
# Count total number of fern fossils in database
n_fern_fossils_total <- 
	fossil_ferns_raw %>%
	# Exclude non-fern fossils
	filter(!affinities %in% c("Euphyllophytes", "Tracheophytes")) %>%
	pull(fossil_taxon) %>%
	n_distinct() %>%
	number()

# Format fossils for printing in MS
fossils <-
	fossil_ferns_raw %>%
	mutate(age = number(minimum_age, accuracy = 0.1)) %>%
  split(.$affinities)

# Format values used for treePL smoothing as tibble
cv_smooth_test <-	
	tibble(cv_result = treepl_cv_results) %>%
	mutate(smooth = str_match(cv_result, "\\((.*)\\)") %>% 
				 	magrittr::extract(, 2) %>%
				 	parse_number()) %>%
	mutate(chisq = str_match(cv_result, "\\) (.*)$") %>% 
				 	magrittr::extract(, 2) %>%
				 	parse_number()) %>%
	arrange(smooth)

# Extract start, end, and best smooth values
smooth_start <- cv_smooth_test %>%
	slice_min(n = 1, order_by = smooth) %>%
	pull(smooth) %>%
	number(., accuracy = .)

smooth_end <- cv_smooth_test %>%
	slice_max(n = 1, order_by = smooth) %>%
	pull(smooth) %>%
	number(., accuracy = .)

smooth_best <-
	cv_smooth_test %>%
	slice_min(n = 1, order_by = chisq) %>%
	pull(smooth) %>%
	number(., accuracy = .)
```

We rooted the ML tree on bryophytes and estimated divergence dates using penalized likelihood with treePL [@Smith2012].
We selected `r n_calibrations_ferns` fern fossils to use as calibration points from a curated list of `r n_fern_fossils_total` fern fossil taxa (`r s_table("fossils")`).
In the case of redundant fossils (those calibrating the same node in the phylogeny; `r eg`, stems of families that are sister to each other, such as Dipteridaceae and Matoniaceae) we selected the fossil with the oldest age (`r ie`, the upper limit of the oldest stratigraphic period assigned to the fossil). 
We assigned fossils to lineages only after consulting the original publication (as opposed to only using the fossil species or genus) so we could reassess the identification relative to changes in taxonomy and hypotheses of phylogenetic relationships.
Taxon concepts may vary between the original publication and currently applied taxonomy (`r eg`, Dicksoniaceae *sensu lato* including other tree fern families used in description of the fossil but Dicksoniaceae *sensu stricto* used currently) and hypotheses of phylogenetic relationships among extant species may change (`r eg`, *Dennstaedtia* used in description of a fossil but this genus now known to be polyphyletic); in both cases, the resulting node to be calibrated for a given fossil will also change.
We also applied two calibration points outside of ferns including the most recent common ancestor (MRCA) of euphyllophytes (`r fossils$Euphyllophytes$age` Ma) and MRCA of tracheophytes (`r fossils$Tracheophytes$age` Ma), and fixed the root age (land plants) at 475 Ma [@Testo2016a].
All calibration points other than the root are minimum ages.

We tested rate smoothing parameters in treePL ranging from `r smooth_start` to `r smooth_end` (each varies by one order of magnitude); a value of `r smooth_best` was selected with the smallest chi-squared value for final analysis.
To construct confidence intervals for the estimated divergence times, we constructed 100 bootstrap trees (trees with identical topology but different branchlengths) using IQ-TREE and dated each using treePL with the same settings used for the ML tree [@Maurin2020].

## Reproducibility

The workflow is managed in R [@RCoreTeam2020a] with the targets package [@Landau2021].
Code used to generate FTOL is available at https://github.com/joelnitta/ftol.
A docker image to run the code is available at https://hub.docker.com/r/joelnitta/ftol.

# Results

## GenBank mining

```{r genbank-stats}
# Helper function to count number of GenBank accessions in a dataset
count_accs <- function(data) {
	data %>%
	pull(accession) %>%
	n_distinct() %>%
	number()
}

# Manual list of accessions to exclude
n_accs_exclude <-
accs_exclude %>%
	pull(accession) %>%
	n_distinct() %>%
	number()

# Sanger ----

# Number of GenBank accessions resulting from inital query
n_accs_query <- count_accs(fern_sanger_seqs_raw)

# Number of GenBank accessions recovered with superCRUNCH
n_accs_extract <- count_accs(raw_fasta_all)

# Number of GenBank accessions after excluding manual rogues,
# names that couldn't be resolved
n_accs_first_filter <- count_accs(sanger_seqs_combined_filtered)

# Number of GenBank accessions after further excluding
# rogues based on BLAST
n_accs_second_filter <- count_accs(sanger_seqs_rogues_removed)

# Number of GenBank accessions in final selection
# first convert to long form
sanger_accessions_selection_long <-
 sanger_accessions_selection %>%
	select(contains("accession")) %>%
	pivot_longer(values_to = "accession", everything()) %>%
	filter(!is.na(accession))

# Count number of raw accessions in final selection
n_accs_raw_final <- count_accs(sanger_accessions_selection_long)

# Count number of accessions in final selection, split by locus
n_accs_final <- 
	sanger_accessions_selection_long %>%
	# Count accessions containing different loci as distinct
	mutate(
		name = str_remove_all(name, "accession_"),
		accession = paste(name, accession, sep = "_")) %>%
	count_accs()

# Analyze join by method
join_by_count <-
sanger_accessions_selection %>% 
	assert(not_na, join_by, species) %>%
	assert(is_uniq, species) %>%
	tabyl(join_by) %>%
	mutate(join_by = fct_reorder(join_by, n)) %>%
	mutate(
			percent = percent(percent, accuracy = 0.1),
			n = number(n, accuracy = 1)) %>%
	split(.$join_by)
	
# Plastome ----
# Don't include OG in calculations
n_plast_accs_query <-
	plastome_metadata_raw_all %>%
	left_join(
		select(plastome_metadata_renamed, accession, outgroup),
		by = "accession") %>%
	# excluded accessions lack outgroup info, but are all ferns
	mutate(outgroup = replace_na(FALSE)) %>%
	assert(not_na, outgroup) %>%
	filter(outgroup == FALSE) %>%
	count_accs()
n_plast_first_filter <-
	plastome_metadata_renamed %>%
  assert(not_na, outgroup) %>%
	filter(outgroup == FALSE) %>%
	count_accs()
n_accs_plast_final <- 
	plastome_seqs_combined_filtered %>%
	left_join(
		select(plastome_metadata_renamed, accession, outgroup),
		by = "accession") %>%
	assert(not_na, outgroup) %>%
	filter(outgroup == FALSE) %>%
	count_accs()
```

The initial GenBank query for the seven Sanger loci resulted in `r n_accs_query` accessions ("accessions" in this case includes some cases where a single accession contains multiple loci).
Extraction of target loci with superCRUNCH recovered `r n_accs_extract` accessions.
Manual inspection of the initial tree resulted in a list of `r n_accs_exclude` accessions to be excluded as rogues.
`r n_accs_first_filter` accessions were retained after excluding these rogues and species names that could not be resolved to an accepted name in the taxonomic database.
`r n_accs_second_filter` accessions were retained after excluding rogues identified by the all-by-all BLAST search.
The final selection of Sanger loci (one set of concatenated accessions per species or one accession per species if conditions for concatenation not met) included `r n_accs_raw_final` accessions (`r n_accs_final` sequences when counting distinct loci within each accession separately).
The most frequent method for joining accessions across loci was by `r last(join_by_count)$join_by` (`r last(join_by_count)$n` species; `r last(join_by_count)$percent`).
`r join_by_count$unjoined$n` species (`r join_by_count$unjoined$percent`) had accessions that could not be joined or only included one of the seven loci (`r s_table("join_by")`).

The initial GenBank query for fern plastomes resulted in `r n_plast_accs_query` accessions.
`r n_plast_first_filter` accessions were retained after excluding rogues and species names that could not be resolved.
Selection of the longest set of concatenated sequences per species resulted in `r n_accs_plast_final` accessions (species).

## Taxon and locus sampling

```{r sampling-stats, message = FALSE}
# Taxon sampling ----
# Format FTOL sampling as list
tax_coverage <-
	coverage_by_rank %>%
	filter(rank == "major_clade") %>%
	transmute(
		# Be sure to change to factor so list is correct order
		# when splitting
		clade = fct_reorder(name, coverage),
		coverage = percent(coverage, accuracy = 0.1)) %>%
	split(.$clade)

# Format overall FTOL sampling as list
total_sampling <- split(total_coverage, total_coverage$rank)

# Format overall plastome (backbone) sampling as list
bb_total_sampling <- split(bb_total_coverage, bb_total_coverage$rank)

# Locus sampling ----
# Make tibble with one row per locus per species
sanger_locus_sampling <-
sanger_alignment_tbl %>%
	mutate(taxon = map(align_trimmed, rownames)) %>%
	select(locus = target, species = taxon) %>%
	unnest(cols = species) %>%
	# Exclude outgroups
	left_join(
		unique(select(sanger_sampling, species, outgroup)), by = "species") %>%
	filter(outgroup == FALSE) %>%
	select(-outgroup)

# Summarize mean number of loci per species
locus_summ <-
sanger_locus_sampling %>%
	group_by(species) %>%
	summarize(
		n_loci = n_distinct(locus)
	) %>%
	summarize(
		max = max(n_loci) %>% number(accuracy = 1),
		min = min(n_loci) %>% number(accuracy = 1),
		mean = mean(n_loci) %>% number(accuracy = 0.1),
		sd = sd(n_loci) %>% number(accuracy = 0.1)
	)
	
# Format total number of samples per locus as list for MS
locus_total <-
sanger_locus_sampling %>%
	count(locus) %>%
	mutate(
		locus = fct_reorder(locus, n),
		n = number(n)) %>%
	split(.$locus)

# Make tibble with number of species per locus combination
locus_comb_tib <-
sanger_locus_sampling %>%
	group_by(species) %>%
	summarize(comb = paste(sort(locus), collapse = "_")) %>%
	count(comb) %>%
	arrange(desc(n)) %>%
	mutate(
		rbcl_present = str_detect(comb, "rbcL"),
		n_loci = str_count(comb, "_") + 1)

# Format as list for MS
locus_comb <-
	locus_comb_tib %>%
	mutate(
		comb = fct_reorder(comb, n),
		n = number(n, accuracy = 1)) %>%
	split(.$comb)

# Verify the most frequent locus combination is rbcL alone
assertthat::assert_that(
	locus_comb_tib %>%
		filter(n == max(n)) %>%
		pull(comb) == "rbcL")

# Verify that the second most frequent locus combination is all seven species
assertthat::assert_that(
	locus_comb_tib %>%
		arrange(desc(n)) %>%
		slice(2) %>%
		pull(n_loci) == 7)

# Calculate number of species with > 1 locus
sp_mult_loci <-
sanger_locus_sampling %>%
	group_by(species) %>%
	summarize(mult_loci = if_else(n_distinct(locus) > 1, "true", "false")) %>%
	tabyl_fmt(mult_loci) %>%
	split(.$mult_loci)

# Backbone locus sampling ----
bb_locus_tbl <-
plastome_alignment_tbl %>%
	mutate(taxon = map(align_trimmed, rownames)) %>%
	select(locus = target, species = taxon) %>%
	unnest(cols = species) %>%
	group_by(species) %>%
	summarize(n_loci = n_distinct(locus)) %>%
	arrange(desc(n_loci)) %>%
	left_join(
		unique(select(sanger_sampling, species, outgroup)), by = "species") %>%
	filter(outgroup == FALSE) %>%
	select(-outgroup)

bb_locus_summ <-
bb_locus_tbl %>%
	summarize(
		max = max(n_loci) %>% number(accuracy = 1),
		min = min(n_loci) %>% number(accuracy = 1),
		mean = mean(n_loci) %>% number(accuracy = 0.1),
		sd = sd(n_loci) %>% number(accuracy = 0.1)
	)

bb_with_all_loci <-
	bb_locus_tbl %>%
	mutate(all_loci_sampled = if_else(n_loci == 79, "true", "false")) %>%
	tabyl_fmt(all_loci_sampled) %>%
	split(.$all_loci_sampled)
``` 

Taxon sampling for the current version of FTOL includes `r total_sampling$species$coverage` species (`r total_sampling$species$coverage_p`), `r total_sampling$genus$coverage` genera (`r total_sampling$genus$coverage_p`), `r total_sampling$family$coverage` families, and `r total_sampling$order$coverage` orders of ferns (all coverage values are relative to the number of accepted species in Pteridocat).
Coverage varied by major clade from `r tax_coverage %>% first %>% extract2("coverage")` (`r tax_coverage %>% first %>% extract2("clade")`) to `r tax_coverage %>% last %>% extract2("coverage")` (`r tax_coverage %>% last %>% extract2("clade")`) (`r figure("ftol")`). Taxon sampling for the backbone phylogeny (plastome loci) includes `r bb_total_sampling$species$n` species (`r bb_total_sampling$species$coverage_p`), `r bb_total_sampling$genus$coverage` genera (`r bb_total_sampling$genus$coverage_p`), `r bb_total_sampling$family$coverage` families (`r bb_total_sampling$family$coverage_p`), and `r bb_total_sampling$order$coverage` orders of ferns.

The number of fern species sampled per locus in the Sanger dataset ranged from `r first(locus_total)$n` (*`r first(locus_total)$locus`*) to `r last(locus_total)$n`  (*`r last(locus_total)$locus`*).
`r sp_mult_loci$true$n` species (`r sp_mult_loci$true$percent`) were sampled for more than one locus.
Mean `r locus_summ$mean` ± `r locus_summ$sd` loci were sampled per species (all errors standard deviation unless otherwise mentioned).
The most frequent type of locus sampling per species was *rbcL* alone (`r locus_comb$rbcL$n` species) (`r s_figure("locus-upset")`).
The second most frequent type of locus sampling per species was all seven loci together (`r nth(locus_comb, length(locus_comb) - 1)$n` species). 
Locus sampling per species in the plastome dataset ranged from `r bb_locus_summ$min` to `r bb_locus_summ$max` (mean `r bb_locus_summ$mean` ± `r bb_locus_summ$sd` loci per species).
`r bb_with_all_loci$true$n` species (`r bb_with_all_loci$true$percent` of plastome species) included all 79 loci.

## DNA alignments

```{r dna-stats}
# Calculate frequency of missing bases in each dataset overall
plastome_missing_freq <- base_freq_missing(plastome_alignment) %>%
	percent(accuracy = 0.1)
sanger_missing_freq <- base_freq_missing(sanger_alignment) %>%
	percent(accuracy = 0.1)

# Calculate frequency of missing bases by locus for Sanger loci
# Concatenate Sanger loci again here to be sure order is the same
sanger_genes_cat <- do.call(
	ape::cbind.DNAbin,
	c(sanger_alignment_tbl$align_trimmed, fill.with.gaps = TRUE))

sanger_missing_by_gene <-
	# Get start and end position of each gene in concatenated alignment
  sanger_alignment_tbl %>%
	mutate(
		nbp = map_dbl(align_trimmed, ncol),
		end = cumsum(nbp),
		start = end - nbp + 1) %>%
	# Calculate frequency of missing data
	select(gene = target, start, end) %>%
	mutate(
		missing = map2_dbl(
			start, end,
			~base_freq_missing(
				seqs = sanger_genes_cat,
				start = .x,
				end = .y))
	) %>%
	select(-start, -end) %>%
	# Reorder and split for printing
	mutate(
		# Format gene names for printing
		gene = str_replace_all(gene, "-", "--"),
		gene = str_replace_all(gene, "([^-]+)", "*\\1*"),
		gene = fct_reorder(gene, missing),
		missing = percent(missing)) %>%
	split(.$gene)
```

The Sanger DNA alignment was `r number(ncol(sanger_alignment))` bp with `r sanger_missing_freq` missing data (missing bases or gaps) overall; rates of missing data by locus ranged from `r sanger_missing_by_gene %>% first() %>% extract2("missing")` (`r sanger_missing_by_gene %>% first() %>% extract2("gene")`) to `r sanger_missing_by_gene %>% last() %>% extract2("missing")` (`r sanger_missing_by_gene %>% last() %>% extract2("gene")`).
The plastome DNA alignment was `r number(ncol(plastome_alignment))` bp with `r plastome_missing_freq` missing data. 

## Topology

```{r topology-stats}
ferns_bs <- get_bs(plastome_tree, c("Equisetum_arvense", "Danaea_nodosa"))
hym_gle_bs <- get_bs(plastome_tree, c("Dipteris_conjugata", "Hymenophyllum_polyanthos"))
cyath_salv_bs <- get_bs(plastome_tree, c("Plagiogyria_adnata", "Azolla_filiculoides"))
denn_poly_bs <- get_bs(plastome_tree, c("Dennstaedtia_hirsuta", "Pseudophegopteris_aurita"))

fern_monophy_summ <-
fern_monophy_summ_tbl %>%
	mutate(
		n = as.integer(n),
		p = percent(percent)
	) %>%
	split(.$tax_level) %>%
	map(~split(., .$monophyly))
```

Ferns are strongly supported as monophyletic (BS `r ferns_bs`; all subsequent relationships mentioned received BS >=`r percent(.98)` in the backbone phylogeny unless otherwise indicated).
The first split within ferns separates a clade including Equisetidae and Ophioglossidae from all other species (`r figure("backbone")`).
The next split separates Marattiales from the remaining ferns, the leptosporangiates (Polypodiidae).
Within leptosporangiate ferns, Osmundales is sister to all other species.
The next lineage to diverge is a clade including Hymenophyllales and Gleicheniales (BS `r hym_gle_bs`), followed by Schizaeales.
Salviniales was recovered as sister to Cyatheales but without strong support (BS `r cyath_salv_bs`), which are in turn sister to Polypodiales. 

The first split within Polypodiales separates a clade including Saccolomatineae and Lindsaeineae from the remainder of species, followed by the subsequent divergences of Pteridineae, then Dennstaedtiineae, which is sister to the eupolypods with moderate support (BS `r denn_poly_bs`).
There are two major clades within eupolypods, Polypodiineae (eupolypods I) and Aspleniineae (eupolypods II).

Within Polypodiineae, Hypodematiaceae is sister to all remaining species, which diverge successfully in the order Didymochlaenaceae, Dryopteridaceae, Lomariopsidaceae, Nephrolepidaceae, Tectariaceae, Oleandraceae, then finally Polypodiaceae sister to Davalliaceae.

Within Aspleniineae, Cystopteridaceae is sister to the remaining species, which are split into two clades. 
One clade consists of Rhacidosoraceae, Diplaziopsidaceae, Desmophlebiaceae, Hemidictyaceae and Aspleniaceae.
The other clade includes Thelypteridaceae, Woodsiaceae, a clade comprised of Onocleaceae and Blechnaceae, and Athyriaceae.

All sampled orders, suborders, families, and subfamilies were recovered as monophyletic (or monotypic) with the exception of Polypodioideae, which is known to be paraphyletic relative to Grammitidoideae [@PteridophytePhylogenyGroupI2016].
`r fern_monophy_summ$genus$No$n` genera (`r fern_monophy_summ$genus$No$p`) were non-monophyletic (`r s_table("monophy")`).

## Divergence times

We estimate the crown age of ferns to be `r crown_ages$ferns` ma.
Ages of other major crown groups are leptosporangiates (`r crown_ages$leptos` ma), Polypodiales (`r crown_ages$polypodiales` ma), eupolypods I (`r crown_ages$eupoly_i` ma), and eupolypods II (`r crown_ages$eupoly_ii` ma).
Except for eupolypods II, these ages are all older than those reported in the most recent global fern phylogeny [@Testo2016a].
Estimated stem ages of fern families were generally older than previous studies (`r figure("divtimes")`).
We did not compare crown ages of families across studies because differences in crown ages in smaller clades are likely affected by species sampling.

# Discussion

## Nodes of contention in fern phylogeny

```{r phy-discussion-stats}
# Calculate number of genera sampled for Gleicheniales
ppgi_gleich_genera_n <-
	ppgi_taxonomy %>%
	filter(order == "Gleicheniales") %>%
	pull(genus) %>%
	n_distinct()

plastome_gleich_genera_n <-
	bb_coverage_by_rank %>%
	left_join(select(ppgi_taxonomy, name = genus, order), by = "name") %>%
	filter(order == "Gleicheniales") %>%
	pull(name) %>%
	n_distinct()
```

The phylogenetic position of Equisetidae relative to other ferns has long been a point of contention in molecular systematics of ferns.
We recovered Equisetidae as sister to Ophioglossidae (Psilotales + Ophioglossales), in agreement with other plastid phylogenomic analyses [@Grewe2013; @Ruhfel2014; @Gitzendanner2018b; @Lehtonen2019; @Kuo2018].
This contradicts nuclear phylogenomic analyses [@Rothfels2015a; @Qi2018; @Shen2018], most plastid analyses with smaller numbers (`r ca` 3--17) of genes [@Schneider2004d; @Rai2010; @Kuo2011; @Testo2016a] and an analysis including mitochondrial data [@Knie2015], which have recovered Equisetidae as sister to all other ferns; some earlier plastid analyses based on Sanger data also recovered Equisetidae sister to Marattidae with low support [@Pryer2001; @Qiu2006; @Qiu2007].
Aside from alternative resolutions of Equisetidae dependent on nuclear envelope or model parameters [@Wickett2014; @Kuo2018], structural support exists for both Equisetidae as sister to Ophioglossidae [`r ca` 550 bp intron in plastid *rps12i346*\; @Grewe2013] and Equisetidae as sister to all other ferns [`r ca` 70 bp intron in mitochondrial *rpl2*\; @Knie2015].
The clear contradiction between plastid and nuclear phylogenomic data may indicate ancient hybridization or introgression, but robust support for any particular scenario is lacking.

Another enigmatic relationship in ferns is the placement of Hymenophyllales.
The monophyly of Polypodiidae (leptosporangiate ferns) is well supported across many studies and not in doubt, as is the status of Osmundales as the first lineage to diverge from the remainder of leptosporangiates [@Pryer2001; @Schneider2004d; @Schuettpelz2007; @Kuo2011; @Testo2016a].
However, the subsequent placement of Hymenophyllales differs across studies: some recover Hymenophyllales as the next diverging lineage after Osmundales [@Pryer2001; @Testo2016a; @Schneider2004d; @Schuettpelz2006; @Schuettpelz2007], while others recover a clade comprising Hymenophyllales sister to Gleicheniales, which then together are sister to the remaining non-Osmundales leptosporangiates [@Pryer2004a; @Lehtonen2017].
Another possibility supported by recent transcriptomic studies is Hymenophyllaceae sister to Gleicheniaceae, which are in turn sister to Dipteridaceae, resulting in a paraphyletic Gleicheniales [@Qi2018; @Shen2018].
We recovered a monophyletic Gleicheniales sister to Hymenophyllales with moderate support, a relationship that was also observed in some, but not all, analyses of plastome data by @Lehtonen2019 and @Kuo2018.
However, current plastome sampling only includes `r plastome_gleich_genera_n` out of `r ppgi_gleich_genera_n` genera of Gleicheniales; additional taxon sampling may help to resolve this relationship.

Within Polypodiales, the relationship between suborders Pteridineae, Dennstaedtiineae and the eupolypods (Polypodiineae and Aspleniineae) has been difficult to resolve (here designated "P", "D", and "e", respectively).
Most previous plastid studies based on Sanger sequencing have recovered (P, (D, e)) [@Kuo2011; @Schuettpelz2006; @Schuettpelz2007; @Testo2016a; but see @Lehtonen2011].
We recover (D, (P, e)) with moderate support (BS `r denn_poly_bs`); this topology agrees with other phylogenomic studies based on whole plastomes [@Lu2015a; @Lehtonen2019] or nuclear data [@Rothfels2015a; @Qi2018; @Shen2018] as well as a plastid supermatrix [@Lehtonen2011].
Recently the whole plastome study of @Du2021 recovered a novel topology comprising ((D, P), e) under some analysis settings but (P, (D, e)) under others.

Taken together, our results for nodes of contention in the fern phylogeny generally agree with other plastid phylogenomic analyses and are well within the realm of plausible hypotheses generated to date.
We do not consider any of these nodes "solved" by our analysis.
Rather, conclusive resolution apparently still awaits additional sampling and perhaps innovation in phylogenetic methods.

## Revisiting fossil calibration points in ferns

We recovered older crown ages for ferns and other large clades as well as older stem ages for families relative to previous studies (Figures `r figure_num("divtimes")`, `r s_figure_num("du-ages")`).
This may be due to the fact that we used a completely revised set of fossil calibration points relative to previous studies, which not only resulted in a more densely calibrated tree but also a higher number of fern families with minimum fossil ages.
Several of the recent studies analyzing divergence times for global fern phylogenies all used a similar set of 24--26 fossil calibration points [@Schuettpelz2007b; @Testo2016a] or secondary calibration points based on studies using this set [@Rothfels2015a].
However, we found several of these calibration points were not appropriately applied due to differences in taxonomic concepts between the original fossil publication and the phylogeny in which they were used.
Moreover, our set of calibration points more than doubles the number used by previous studies.
It is likely that this expanded sampling has yielded older estimates for the stem age of many families, whereas previous studies tended to estimate similar ages by re-using the same (or very similar), more limited set of calibration points (Figures `r figure_num("divtimes")`, `r s_figure_num("du-ages")`).

<<FIXME: add discussion of some specific groups once we have final ages>>

## Plastid vs. nuclear fern trees

Currently, the vast majority of publicly available DNA sequence data for ferns is Sanger-sequenced plastid gene regions (`r figure("genbank")`).
Plastid sequences are convenient for phylogenetic analysis because they are essentially a single, uniparentally inherited linkage group, thus free from recombination.
However, the plastid tree does not necessarily agree with trees inferred from other data sources.
Conflict between plastid and nuclear phylogenies has been frequently observed in narrowly focused (`r eg`, genus level) studies using traditional Sanger sequencing, and has recently been demonstrated at deeper levels within Polypodiaceae using phylogenomic approaches [@Wei2022].
Such conflict does not necessarily reflect insufficient methodology or sampling, but rather may be due processes including (but not limited to) introgression, lineage sorting, and hybridization at deep phylogenetic levels.
Therefore, we have no expectation that there is a single fern tree of life; rather, multiple trees exist that each capture an element of biological reality, and FTOL is merely one such possible tree.

Recent studies using transcriptomics are clarifying the backbone of the fern phylogeny using many (25--2400) nuclear genes from representative species spanning the tree [@Rothfels2015a; @Qi2018; @Shen2018].
The most comprehensively sampled phylogenomic study targeting ferns is the on-going Genealogy of Flagellate Plants (GoFlag) project, which seeks to generate genomic data (ca. 300 single-to-low copy nuclear gene regions) for all flagellate plants (bryophytes, lycophytes, ferns, and gymnosperms) [@Breinholt2021].
GoFlag data have recently been used in a phylogenomic analysis and taxonomic revision of Thelypteridaceae resulting in multiple new genera [@Fawcett2021; @Fawcett2021a], and additional phylogenomic analyses of other fern groups using GoFlag markers are to be expected in the near future.
The rapid growth of genomic data notwithstanding, species level sampling of such nuclear phylogenomic datasets still is far less that available for plastid data (`r figure("genbank")`).
Furthermore, many subclades of ferns are under active investigation using plastid markers using both Sanger and next-gen sequencing, and plastid data for previously unsampled species will likely continue to grow at a rapid pace.
We therefore expect that the methodology outlined here will continue to be useful to generate a maximally sampled plastid fern tree of life for several more years.

## Accesibility and usage of FTOL

We have sought to make FTOL easily available to support research on the evolution and ecology of ferns.
FTOL is available via the "ftol" R package, as well as a web portal (FIXME: add URL when ready).
Furthermore, we have made all the underlying data (`r eg`, DNA alignments, fossil calibration points) available so that other researchers can use these to conduct analyses such as further investigations of divergence times or phylogenetic analysis including custom DNA sequences [`r eg`\; @Nitta2021.08.26.457744].

A typical step in any analysis that joins data across multiple sources (`r eg`, trait data and a phylogeny) is to resolve taxonomic names so that the usage of synonyms does not prevent data from joining [@Page2008].
During the preparation of FTOL, we developed two additional R packages that enable taxonomic name resolution to join data with FTOL: the "pteridocat" package (https://github.com/joelnitta/pteridocat) and the "taxastand" package (https://github.com/joelnitta/taxastand).
We selected R because it is widely used by the ecological research community, well established, and freely available [@Lai2019].
The "pteridocat" package includes the pteridocat taxonomic database as a data frame (tibble) in Darwin Core format.
The "taxastand" package includes functions to resolve taxonomic names while taking account variation in taxonomic author format and orthographic variation.
By using these two packages in combination, it should be straightforward for other researchers to map their own data onto FTOL, thus greatly enabling and enhancing studies including, but not limited to, comparative phylogenetics, biogeography, and community ecology in ferns.

## FTOL as a living, community-driven resource

We want to be clear that FTOL is no way meant to be the "official" fern phylogeny; it is simply one reasonable hypothesis that has been designed to be maximally inclusive at the species level.
FTOL cannot substitute for systematic careful studies at finer taxonomic scales that include sampling multiple individuals per species and/or other sources (`r eg`, morphological, nuclear) of data, and such studies continue to be vital to our understanding of fern evolution.

One feature of FTOL that sets it apart from the vast majority of other fern phylogenetic studies is its iterative nature.
Unlike most other published fern phylogenies, the current version of FTOL described in this manuscript is not meant to be the last.
Rather, we plan to re-run these analyses as additional data become available on GenBank, and release updated versions of the tree on a regular, semi-annual basis.
We envision that FTOL will be integrated with the next iteration of PPG, PPGII, to provide the most recent hypotheses on the monophyly of various fern taxa, which will in turn enable a more natural classification system.

Furthermore, FTOL will not only grow in size with time, but also become more refined.
We are aware that our methodology cannot produce a "perfect" tree, nor is that our goal.
Indeed, we anticipate that there will almost always be tips in the tree that need correction, either because they get overlooked (`r eg`, placement of species within genera, which we did not have the resources to inspect), or proper taxonomic treatment is not yet available (`r eg`, cheilanthoid ferns).
That is why we have made our methodology (code), data, and software (R packages and docker image) completely open and available to the research community.
It is our hope that other researchers using FTOL will contribute by making edits and suggestions, preferably through the GitHub repository (https://github.com/joelnitta/ftol).
This way, FTOL will continually improve and keep pace with the currently available data and taxonomic hypotheses of ferns.

# Acknowledgments {-}

Members of the Iwasaki lab provided helpful comments that improved the manuscript.
This study was supported by JSPS KAKENHI Grant Number 16H06279.
The authors thank Michael Hassler for maintaining the World Ferns taxonomic database and making it available to use for research.

# References {-}
<!-- Insert <div id="refs"></div> to place the refs here instead of at the end -->
<div id="refs"></div>

`r pagebreak_pdf(params)`

# Figures {-}

```{r plot-prep}
### plot-prep start

# Make vector of outgroup taxa
og_taxa <-
	sanger_sampling %>%
	filter(outgroup == TRUE) %>%
	pull(species)

plastome_tree_rooted <- phytools::reroot(
	plastome_tree,
	getMRCA(
		plastome_tree,
		c("Physcomitrium_patens", "Marchantia_polymorpha", "Anthoceros_angustus")
	)) %>%
	ladderize()

# Make tibble of backbone tips with higher-level taxonomy
backbone_taxonomy <-
	tibble(species = get_tips_in_ape_plot_order(plastome_tree_rooted)) %>%
	# Add higher taxonomy
	left_join(sanger_sampling, by = "species") %>%
	# Add column for phylogenetic order
	mutate(phy_order = seq_len(nrow(.))) %>%
	# Outgroup lacks family data, so call "og"
	mutate(
		family = case_when(
			outgroup == TRUE ~ "og",
			TRUE ~ family),
		major_clade = case_when(
			outgroup == TRUE ~ "og",
			TRUE ~ major_clade)
	)

# Make exemplar set of species in backbone representing plot groups
major_clade_exemp_taxa <-
	backbone_taxonomy %>%
	# Slice to one exemplar species per major clade
	group_by(major_clade) %>%
	slice(1) %>%
	ungroup() %>%
	# Slicing arranges alphabetically, so put back in (unladderized) phylo order
	arrange(phy_order) %>%
	# Drop outgroup
	filter(family != "og")

# Make a color palette for major clades
# Start with list of green.armytage.colors names,
# exclude those that are too close or stand out too much
# https://www.jstatsoft.org/article/view/v090c01
green_army_use_cols <-
	c(
		"amethyst",
		"blue",
		"caramel",
		"damson",
		# "ebony",
		# "forest",
		"green",
		"honeydew",
		# "iron",
		# "jade",
		# "khaki",
		# "lime",
		"mallow",
		"navy",
		# "orpiment",
		"pink",
		# "quagmire",
		"red",
		"sky",
		"turquoise",
		"uranium",
		"violet",
		"wine",
		# "xanthin",
		# "yellow",
		"zinnia"
	)

clade_cols <-
	Polychrome::green.armytage.colors() %>%
	extract(green_army_use_cols) %>%
	extract(seq_len(nrow(major_clade_exemp_taxa))) %>%
	set_names(major_clade_exemp_taxa$major_clade)

clade_cols_tib <-
	tibble(
		color = clade_cols,
		major_clade = names(clade_cols)
	) %>%
	mutate(
		code = substr(major_clade, 1, 2),
		code = case_when(
			major_clade == "Saccolomatineae" ~ "Sac",
			major_clade == "Salviniales" ~ "Sal",
			TRUE ~ code
		)) %>%
	assert(is_uniq, code)

# Make tibble for annotating major clades
clade_tibble <- 
	# Use assess_monophy to get MRCA of each major clade
	# They should all by monophyletic anyways
	select(sanger_sampling, species, major_clade) %>%
	assess_monophy(
		taxon_sampling = .,
		tree = plastid_tree_dated,
		tax_levels = "major_clade"
	) %>%
	get_result_monophy(1) %>%
	# Exclude lycophytes
	filter(!taxon %in% c(
		"Selaginellales", "Lycopodiales", "Isoëtales", "Isoetales")) %>%
	verify(all(monophyly == "Yes")) %>%
	transmute(major_clade = taxon, mrca = parse_number(mrca)) %>%
	left_join(clade_cols_tib, by = "major_clade")
### plot-prep end
```

```{r genbank-make}
fern_tree_sampling <-
	tribble(
		~source, ~year, ~n_species, ~label,
		"Hasebe1995", 1995, 107, "Hasebe et al. 1995",
		"Schuettpelz2007", 2007, 400, "Schuettpelz et al. 2007",
		"Lehtonen2011", 2011, 2957, "Lehtonen 2011",
		"Testo2016a", 2016, 3973, "Testo and Sundue 2016",
		"this_study", 2021, n_sanger_ferns, "This study",
	) %>%
	mutate(source = fct_reorder(source, year))

compartment_labs <- 
	gb_species_by_year %>%
	filter(year == max(year)) %>%
	transmute(
		type,
		label = str_to_sentence(type),
		year,
		n_species
	)

okabe_ito_cols <-
	palette_okabe_ito(1:9) %>%
	set_names(
		c("orange", "light_blue", "green", 
			"yellow", "blue", "red",
			"magenta", "grey", "black"))

compartment_cols <- c(
	mitochondrial = okabe_ito_cols["red"] %>% set_names(NULL),
	nuclear = okabe_ito_cols["blue"] %>% set_names(NULL),
	plastid = okabe_ito_cols["green"] %>% set_names(NULL),
	total = "grey40"
)

# Manually specify axis buffers since we will annotate the panel border manually
year_buffer <- 
	max(gb_species_by_year$year) %>%
	subtract(min(gb_species_by_year$year)) %>%
	multiply_by(0.05)
sp_buffer <- max(gb_species_by_year$n_species) %>%
	subtract(min(gb_species_by_year$n_species)) %>%
	multiply_by(0.05)

genbank_plot <-
	ggplot(gb_species_by_year, aes(x = year, y = n_species)) +
	geom_line(aes(color = type)) +
	scale_color_manual(values = compartment_cols, guide = "none") +
	# Annotate panel border manually, since the automatic border gets
	# printed on top of label annotations with theme_bw()
	annotate(
		"rect", 
		xmin = min(gb_species_by_year$year) - year_buffer,
		xmax = max(gb_species_by_year$year) + year_buffer,
		ymin = min(gb_species_by_year$n_species) - sp_buffer,
		ymax = max(gb_species_by_year$n_species) + sp_buffer,
		fill = "transparent",
		color = "grey20") +
	geom_label_repel(
		data = compartment_labs %>% filter(type %in% c("nuclear", "mitochondrial")),
		aes(color = type, label = label),
		segment.linetype = 2, 
		xlim = c(2020, Inf),
		# Do not repel from top or bottom edges.
		ylim = c(-Inf, Inf)
	) +
	geom_label_repel(
		data = compartment_labs %>% filter(type %in% c("total", "plastid")),
		aes(color = type, label = label),
		segment.linetype = 2, 
		xlim = c(2020, Inf),
		ylim = c(-Inf, Inf),
		min.segment.length = 0
	) +
	geom_point(
		data = fern_tree_sampling,
		aes(shape = source)
	) +
	scale_shape(
		limits = fern_tree_sampling$source, 
		labels = fern_tree_sampling$label
	) +
	scale_x_continuous(expand = expansion(0, 0)) + 
	scale_y_continuous(expand = expansion(0, 0)) + 
	labs(shape = "Study", y = "No. species", x = "Year") +
	coord_cartesian(clip = "off") +
	guides(shape = guide_legend(nrow = 2)) +
	theme_bw() +
	theme(
		panel.border = element_blank(),
		legend.position = "bottom",
		legend.title = element_blank(),
		plot.margin = margin(r = 1, t = 0.5, l = 0.5, unit = "in")
	)
```

```{r genbank-save, eval = params$doc_type == "doc" }
# Save plot if rendering to doc
ggsave(
	result_file(figure("genbank"), "pdf"),
	genbank_plot, width = 7, height = 5, units = "in")
```

```{r genbank-print, fig.width = 7, fig.height = 5, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
genbank_plot
```

**`r figure("genbank")`**.
Number of fern species in GenBank by year and genomic compartment.
Points indicate number of species sampled in selected studies of global fern phylogeny [@Hasebe1995; @Schuettpelz2007; @Lehtonen2011; @Testo2016a\; this study].
@Schuettpelz2007 did not attempt exhaustive sampling but rather proportional sampling according to lineage size.
The relatively small increase in number of species in 2021 is likely due to accessions that are still embargoed at the time of writing.
Taxonomy of GenBank species follows NCBI [@Federhen2012].
Only accessions identified to species included; environmental samples, hybrid formulas, and names with "aff." or "cf." annotations excluded.

`r pagebreak_pdf(params)`

```{r ftol-make}
# Coverage bar plot ----

# Format y-axis text: show major clade abbreviations in bold
clade_labs_tib <-
	clade_cols_tib %>%
	select(major_clade, code) %>%
	mutate(
		clade_short = map2_chr(major_clade, code, ~str_remove(.x, .y)),
		label = glue("<b>{code}</b>{clade_short}"))

clade_labs <- clade_labs_tib$label %>%
	set_names(., clade_labs_tib$major_clade)
	
# Format coverage at order-ish level for plotting
b <-
	coverage_by_rank %>%
	filter(rank == "major_clade") %>%
	left_join(major_clade_exemp_taxa, by = c(name = "major_clade")) %>%
	arrange(phy_order) %>%
	rename(clade = name) %>%
	mutate(
		clade = fct_reorder(clade, phy_order) %>%
			fct_rev(),
		coverage_text = glue("{number(n, accuracy = 1)}/{number(n_accepted, accuracy = 1)}")
	) %>%
	ggplot(aes(x = coverage, y = clade, fill = clade)) +
	geom_col(color = "grey10", size = 0.2) +
	geom_text(
		aes(label = coverage_text),
		hjust = 0,
		nudge_x = 0.01,
		size = 6*.36 # 6 pt font
	) +
	scale_x_continuous(
		labels = percent,
		limits = c(0,1),
		expand = expansion(mult = c(0, 0))
	) +
	scale_fill_manual(values = clade_cols) +
	scale_y_discrete(labels = clade_labs) +
	labs(x = "Coverage") +
	theme_bw() +
	theme(
		axis.text.y = element_markdown(size = 6),
		axis.title.y = element_blank(),
		legend.position = "none",
		plot.background = element_rect(fill = "transparent", colour = NA),
		panel.grid.major.y = element_blank(),
		panel.grid.minor.x = element_blank()
	)

# Tree plot ----

# Helper function to label clades with different offset
# For some reason, using the same offset.text around the plot
# results in different positions of text relative to bars. 
# Account for this by manual adjusting offset.text
ftol_geom_cladelab <- function(code_select, offset_text) {
	geom_cladelab(
		data = filter(clade_tibble, code %in% code_select),
		mapping = aes(
			node = mrca,
			color = major_clade,
			label = code
		),
		offset = 0.1,
		offset.text = offset_text,
		fontsize = 2.5,
		textcolour = "black",
		barsize = 2)
}

a <- ggtree(
	plastid_tree_dated, layout = "circular", right = FALSE, size = 0.1) + 
	ftol_geom_cladelab(
		clade_tibble$code[clade_tibble$code %in% c("Eq", "Op", "Ma", "Os", "Gl")], 10) +
	ftol_geom_cladelab(
		clade_tibble$code[clade_tibble$code %in% c("Hy", "Sc", "Sal")], 14) +
	ftol_geom_cladelab(
		clade_tibble$code[clade_tibble$code %in% c("Cy", "Sac", "Li", "De")], 17) +
	ftol_geom_cladelab("Pt", 26) +
	ftol_geom_cladelab("As", 34) +
	ftol_geom_cladelab("Po", 17) +
	scale_color_manual(values = clade_cols) +
	theme(legend.position = "none")
	
# Assemble plots ----
ftol_plot <-
	a + inset_element(b, left = 0.85, bottom = 0.14, right = 1.3, top = .8) +
	plot_annotation(
		theme = theme(
			plot.margin = unit(c(-2,5,-2,-1.4), "cm"))
	)
```

```{r ftol-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
	result_file(figure("ftol"), "pdf"),
	ftol_plot, width = 7, height = 7, units = "in")
```

```{r ftol-print, fig.width = 7, fig.height = 7, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
ftol_plot
```

**`r figure("ftol")`**.
Fern Tree of Life (FTOL).
Tree rooted on bryophytes.
Inset plot shows coverage by major clade (order or suborder).
Bold part of each clade name is its code, which is also indicated on the tree.
Numbers next to each bar show sampled species out of total number of species.
Taxonomy follows @PteridophytePhylogenyGroupI2016.

`r pagebreak_pdf(params)`

```{r backbone-make}
# Make exemplar set of species in backbone representing families
family_exemp <-
	backbone_taxonomy %>%
	# Drop outgroups %>%
	filter(family != "og") %>%
	# Slice to one exemplar species per family
	group_by(family) %>%
	slice(1) %>%
	ungroup() %>%
	# Slicing arranges alphabetically, so put back in (unladderized) phylo order
	arrange(phy_order)

# Trim tree to exemplar species, rename by family
phy <- keep.tip(plastome_tree_rooted, family_exemp$species)
new_tip_label <-
	tibble(species = phy$tip.label) %>%
	left_join(family_exemp, by = "species") %>%
	pull(family)
phy$tip.label <- new_tip_label
# Convert BS support values to numeric
phy$node.label <- parse_number(phy$node.label)
# Scale to total height of 100 for easy plotting
phy <- rescale_tree(phy, 100)

# Make tibble for printing BS values
bs_tibble <- tibble(
	# hard-code node ID: internal nodes start after tip nodes,
	# and phy$node.label is in the same order as internal nodes
	node = 1:Nnode(phy) + Ntip(phy), 
	# Don't print BS = 100%
	bootstrap = ifelse(phy$node.label == 100, "", phy$node.label))

# Make tibble for labeling clades
phy_tip <- as_tibble(phy)

suborder_labels_singletons <-
	family_exemp %>% 
	filter(outgroup == FALSE) %>%
	group_by(major_clade) %>%
	select(major_clade, family) %>% 
	filter(n_distinct(family) == 1) %>%
	ungroup() %>%
	left_join(phy_tip, by = c(family = "label")) %>%
	select(major_clade, mrca = node)

suborder_labels <-
	family_exemp %>% 
	filter(outgroup == FALSE) %>%
	group_by(major_clade) %>%
	select(major_clade, family) %>% 
	filter(n_distinct(family) > 1) %>%
	summarize(taxa = list(family)) %>%
	mutate(
		mrca = map_chr(taxa, ~getMRCA(phy, .))
	) %>%
	transmute(
		major_clade,
		mrca
	) %>%
	mutate(mrca = parse_number(mrca)) %>%
	bind_rows(suborder_labels_singletons) %>%
	mutate(
		clade_label = case_when(
			major_clade == "Aspleniineae" ~ "Aspleniineae\n(eupolypods II)",
			major_clade == "Polypodiineae" ~ "Polypodiineae\n(eupolypods I)",
			TRUE ~ major_clade
		)
	)

# Get data of x and y co-ordinates for ggtree plot
phy_plot_data <- ggplot_build(ggtree(phy)) %>%
	magrittr::extract2("data") %>%
	magrittr::extract2(1) %>%
	as_tibble()

# Make a dataframe for plotting various clade labels
other_clade_labels <-
	list(
		"eupolypods" = c("Polypodiaceae", "Aspleniaceae"),
		"Polypodiales" = c("Polypodiaceae", "Saccolomataceae"),
		"Polypodiidae\n(leptosporangiates)" = c("Polypodiaceae", "Osmundaceae")
	) %>%
	tibble(
		clade_label = names(.),
		x = map_dbl(., ~node_height_from_tips(phy, .)),
		node = map_dbl(., ~ape::getMRCA(phy, .))
	) %>%
	left_join(
		select(phy_plot_data, node, y), by = "node"
	) %>%
	select(-`.`, -node)

segment_cols <-
	rep("grey20", length(other_clade_labels$clade_label)) %>%
	set_names(other_clade_labels$clade_label)

set.seed(123)
# Use the ggtree::`%<+%` operator to map the 
# bootstrap values onto the ggtree object
backbone_plot <- 
	ggtree(phy) %<+% bs_tibble +
	# now we can show BS values using geom_text()
	geom_text(aes(label = bootstrap), hjust = -.25, size = 2.5) +
	# add tip labels
	geom_tiplab(aes(label = label)) +
	# highlight major clades
	geom_hilight(
		data = suborder_labels,
		mapping = aes(
			node = mrca,
			fill = major_clade
		),
		alpha = 0.2,
		align = "right",
		extend = 30,
		gradient.direction = "rt",
		roundrect = TRUE,
		gradient.length.out = 2) +
	scale_fill_manual(values = clade_cols) +
	# label major clades
	geom_cladelab(
		data = suborder_labels,
		mapping = aes(
			node = mrca,
			label = clade_label
		),
		offset = 28,
		align = TRUE,
		extend = 0.3,
		hjust = 0,
		barcolour = "grey20",
		barsize = 2) +
	# Label other groups
	geom_text_repel(
		data = other_clade_labels %>%
			filter(clade_label == "Polypodiidae\n(leptosporangiates)"),
		aes(x = x, y = y, label = clade_label),
		hjust = 0,
		xlim = c(NA, 10),
		ylim = c(15, NA),
		color = "grey20",
		segment.color = "grey20"
	) +
	geom_text_repel(
		data = other_clade_labels %>%
			filter(clade_label == "eupolypods"),
		aes(x = x, y = y, label = clade_label),
		hjust = 0,
		xlim = c(NA, 64),
		ylim = c(32, NA),
		color = "grey20",
		segment.color = "grey20"
	) +
	geom_text_repel(
		data = other_clade_labels %>%
			filter(clade_label == "Polypodiales"),
		aes(x = x, y = y, label = clade_label),
		hjust = 0,
		xlim = c(NA, 48),
		ylim = c(24, NA),
		color = "grey20",
		segment.color = "grey20"
	) +
	xlim(0, 180) +
	coord_cartesian(clip = "off") +
	theme(legend.position = "none")
```

```{r backbone-tree-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
	result_file(figure("backbone"), "pdf"),
	backbone_plot, width = 7, height = 9, units = "in")
```

```{r backbone-tree-print, fig.width = 7, fig.height = 9, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
backbone_plot
```

**`r figure("backbone")`**.
Fern tree of life (FTOL) backbone phylogeny.
One exemplar tip is shown per family.
Ultrafast bootstrap support values (%) shown at nodes; unlabeled nodes are 100%.
Outgroup (seed plants, lycophytes, and bryophytes) not shown.
Colors of major clades (orders or suborders) correspond to those used in `r figure("ftol")`. Taxonomy follows @PteridophytePhylogenyGroupI2016; informal clade names in lowercase.

`r pagebreak_pdf(params)`

```{r div-times-make}
# Convert family ages to almost-long format

# Helper function for subsetting dates to get into
# almost-long format
subset_ages <- function(other_dates, author_sel) {
	other_dates %>%
		select(family, contains(glue("{author_sel}_"))) %>%
		rename_with(~str_remove_all(., glue("{author_sel}_"))) %>%
		mutate(author = author_sel)
}

# Combine dates in almost-long format:
# one row per family/study combination,
# cols for median, high, and low date
all_ages <-
bind_rows(
	subset_ages(other_dates, "ts"),
	subset_ages(other_dates, "rothfels"),
	select(other_dates, family, median = schuettpelz_best) %>%
		mutate(author = "schuettpelz"),
	transmute(
		family_stem_ages,
		family,
		author = "current",
		median = age
	)
) %>%
	left_join(
		unique(select(sanger_sampling, family, major_clade)), by = "family") %>%
	left_join(clade_tibble, by = "major_clade") %>%
	# (some median dates are NA if that family wasn't included in a study)
	assert(not_na, family, major_clade, code, author) %>%
	# Set family as a factor ordered by current age estimate
	left_join(family_stem_ages, by = "family") %>%
	mutate(family = fct_reorder(family, age)) %>%
	select(-age)
 
# Define shapes and colors for points
ages_shapes <- c(
	current = 16, # circle
	rothfels = 17, # triangle
	schuettpelz = 15, # square
	ts = 18 # diamond
)

ages_colors <- c(
	current = palette_okabe_ito(order = "vermillion"),
	rothfels = palette_okabe_ito(order = "bluishgreen"),
	schuettpelz = palette_okabe_ito(order = "blue"),
	ts = palette_okabe_ito(order = "orange")
)

# Format names of papers for printing
ages_sampling <-
	tribble(
		~source, ~year, ~label,
		"schuettpelz", 2009, "Schuettpelz & Pryer 2009",
		"rothfels", 2015, "Rothfels et al. 2015",
		"ts", 2016, "Testo and Sundue 2016",
		"current", 2021, "This study",
	) %>%
	mutate(source = fct_reorder(source, year))
	
# Format y-axis text: show major clade in parentheses
family_labs_tbl <-
all_ages %>%
	select(family, code) %>%
	unique() %>%
	mutate(label = glue("{family} ({code})"))

family_labs <- family_labs_tbl$label %>%
	set_names(., family_labs_tbl$family)
	
# Provide names of periods in caption
# Make sure to read and check that they are all there!
period_names <-
	periods %>%
	# Filter to range including Ordovician to Neogene
	filter(min_age < 475) %>%
	filter(min_age > 1) %>%
	arrange(desc(min_age)) %>%
	mutate(label = glue("{abbr} ({name})")) %>%
	pull(label) %>%
	paste(collapse = ", ")

div_times_plot <-
ggplot(all_ages) +
	geom_pointrange(
		aes(
			x = median, xmin = low, xmax = high,
			y = family,
			color = author, shape = author),
		position = position_dodge(width = 0.5)) +
	scale_shape_manual(
		values = ages_shapes,
		limits = ages_sampling$source, 
		labels = ages_sampling$label) +
	scale_color_manual(
		values = ages_colors,
		limits = ages_sampling$source, 
		labels = ages_sampling$label) +
	scale_y_discrete(labels = family_labs, expand = expansion(add = 0.7)) +
	scale_x_reverse("Age (millions of years)") +
	coord_geo(
		dat = list("periods", "eras"),
		pos = list("b", "b"),
		abbrv = list(TRUE, FALSE),
		xlim = c(500, 0),
		fill = "transparent",
		color = "grey20",
		lab_color = "grey30",
		height = unit(1, "lines"),
		# FIXME: change this to true when issue gets resolved
		# https://github.com/willgearty/deeptime/issues/41
		expand = FALSE,
		size = 3) +
	guides(shape = guide_legend(nrow = 2)) +
	theme_bw() +
	theme(
		axis.title.y = element_blank(),
		legend.title = element_blank(),
		legend.position = "bottom"
	)
```

```{r div-times-save, eval = params$doc_type == "doc"}
# Save plot if rendering to doc
ggsave(
	result_file(figure("divtimes"), "pdf"),
	div_times_plot, width = 7, height = 8, units = "in")
```

```{r div-times-print, fig.width = 7, fig.height = 9, eval = params$doc_type == "pdf"}
# Print plot if rendering to PDF
div_times_plot
```

**`r figure("divtimes")`**.
Stem age of fern families (Ma) estimated by selected studies.
For studies that used methods with confidence intervals, error bars indicate lower and upper 95% highest posterior density levels and point indicates median [@Rothfels2015a; @Testo2016a].
For other studies, point indicates best (most likely) estimate [@Schuettpelz2007b\; this study].
Codes in parentheses after family names indicate major clade as in `r figure("ftol")`.
Period name abbreviations as follows: `r period_names`.