---
title: "An open and continuously updated fern tree of life (FTOL)"
output: 
  bookdown::word_document2:
    pandoc_args: [ "--csl", "frontiers.csl"]
    toc: no
    number_sections: no
    reference_docx: Frontiers_Template.docx
editor_options: 
  chunk_output_type: console
---

---
bibliography: 
- `r here::here("ms/references.yaml")`
- `r here::here("ms/references_other.yaml")`
- `r here::here("ms/references_other.bib")`
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(targets)
library(assertr)
library(ape)
library(scales)
library(tidyverse)
```

```{r load-targets}
tar_load(
	c(
		sanger_alignment,
		plastome_alignment,
		plastome_tree,
		plastome_metadata_renamed,
		ppgi_taxonomy),
	store = here::here("ftol_cache")
)
```

```{r abstract-stats}
# Make tibble of outgroup species
og_species <- 
	plastome_metadata_renamed %>%
	select(species, outgroup) %>%
	filter(outgroup == TRUE)

# Make tibble with one row per species in Sanger sampling
sanger_sampling <- tibble(
	species = rownames(sanger_alignment)
) %>%
	# Add higher-level taxonomy
	mutate(genus = str_split(species, "_") %>% map_chr(1)) %>%
	left_join(
		select(ppgi_taxonomy, genus, family, order), by = "genus"
	) %>% 
	# Add outgroup status
	left_join(og_species, by = "species") %>%
	mutate(outgroup = replace_na(outgroup, FALSE)) %>%
	verify(nrow(.) == nrow(sanger_alignment)) %>%
	verify(sum(outgroup) == nrow(og_species))

# Count number of fern species in Sanger data
n_sanger_ferns <-
	sanger_sampling %>% 
	filter(outgroup == FALSE) %>%
	n_distinct(.$species)
```


<!-- Set left justification -->
\raggedright

<!-- Don't indent -->
\setlength{\parindent}{0ex}

**Joel H. Nitta^1^, Eric Schuettpelz^2^, Santiago Ramírez-Barahona^3^, and Wataru Iwasaki^4^**

^1^Department of Biological Sciences, Graduate School of Science, The University of Tokyo, Tokyo, Japan

^2^Department of Botany, National Museum of Natural History, Smithsonian Institution, Washington, D.C., USA

^3^Departamento de Botánica, Instituto de Biología, Universidad Nacional Autónoma de México, Mexico City, Mexico

^4^Department of Integrated Biosciences, Graduate School of Frontier Sciences, The University of Tokyo, Chiba, Japan

**\*Correspondence:**    
Joel H. Nitta  
joelnitta@gmail.com

Keywords: Fern~1~, Phylogeney~2~, Plastome~3~, PPGI~4~, Pteridophyte~5~, *rbcL*~6~.

# Abstract

Ferns (ca. 12,000 spp.) are the second most diverse clade of vascular plants after seed plants and have received relatively intense focus from molecular systematists, resulting in the publication of molecular trees for nearly all major clades and DNA sequences for ca. 1/3 of all fern species deposited in GenBank.
Global trees including all ferns have been published periodically, but as molecular systematics research continues at a rapid pace, such trees quickly become outdated.
While some large projects exist to generate trees for all plants or all life, such "one-size-fits-all" approaches are not expected to generate optimal results for specific groups because they must work across a very wide range of taxa.
Here, we develop a mostly automated, reproducible, open pipeline to generate a continuously updated fern tree of life (FTOL) from sequence data on GenBank.
Our tailored sampling strategy combines whole plastomes (few taxa, many loci) with Sanger-sequenced loci (many taxa, few loci) to obtain a global fern phylogeny with high resolution along the backbone and maximum sampling across the tips.
We provide multiple versions of the phylogeny including a time-calibrated tree and a synthetic tree with species lacking molecular data added by taxonomy.
We use an expert-curated reference taxonomy to resolve synonyms in compliance with Pteridophyte Phylogeny Group I (PPGI), the community-driven taxonomic system widely accepted by many pteridologists.
The current FTOL (v0.0.1) includes `r comma(n_sanger_ferns)` species, an increase of > 20% relative to the most recent global fern phylogeny.
FTOL will be updated on a semi-annual basis and will be available via a web portal and R package, enabling researchers to have easy access to the most recent, comprehensively sampled fern phylogeny.
FTOL will be useful for anybody studying this important group of plants at scales from smaller (e.g., genus or lower) clades to the entire tree; we anticipate FTOL will be particularly relevant for macroecological studies at the regional to global scale.
Finally, we envision FTOL will be integrated into the next PPG, PPGII, to produce a "living" taxonomic system with direct linkage between a community-driven fern taxonomy and the most recent hypothesis of phylogeny.

# Introduction

Ferns (ca. 12,000 spp.) are the second most diverse clade of vascular plants after seed plants (ca. 260,000 spp.), and are a useful study system for understanding processes of biogeography [@Tryon1986; @Kato1993], community ecology [e.g., @Hennequin2014; @Lehtonen2015], and speciation [e.g., @Kao2020].
As Dobzhansky famously said, "Nothing in biology makes sense except in the light of evolution"; likewise, a well-sampled molecular phylogeny is key to any investigation of fern evolutionary biology.
Fortunately, ferns have received relatively intense focus from molecular systematists, resulting in the publication of molecular trees for nearly all major clades and DNA sequences for ca. 1/3 of all fern species deposited in GenBank.
Thus, there is both sufficient sampling and a pressing need for a globally sampled fern phylogeny.

Past efforts to construct such a global phylogeny have steadily increased in number of species as sequences in GenBank have accumulated [@Pryer2004a; @Lehtonen2011; @Schuettpelz2007; @Testo2016a].
Indeed, accumulation of plastid fern accessions in GenBank show no sign of slowing since the most recent global phylogeny [@Testo2016a].
We believe there is a need therefore, not only for a global fern phylogeny, but one that is self-updating to keep pace with the continued accumulation of molecular data.
This will eliminate the need for researchers to "rebuild the wheel" each time they want to use a globally sampled fern phylogeny.

We recognize that multiple efforts exist to automatically or semi-automatically generate trees for any arbitrary part of the tree of life [@Antonelli2016], a tree for all plants [@Eiserhardt2018], or even the entire tree of life at once [@Hinchliff2015] which would subsume a global fern phylogeny.
While these may be appropriate in some cases, we contend that they cannot produce an optimal fern phylogeny due to the need to take "one-size-fits-all" approaches to accommodate such a wide phylogenetic breadth. 
By focusing methods and datasets specifically on ferns, it should be possible to generate a higher quality end-product (tree) that can be used "as-is" by biologists studying these organisms.

Furthermore, there is much to be gained from integrating a carefully designed global fern phylogeny with the fern systematics community that would not be possible with a "tree of all life" or "tree of all plants".
Recently, a genus-level taxonomy of ferns and lycophytes was established following a open, community-driven model [@PteridophytePhylogenyGroupI2016].
Thanks to this community-first approach, the taxonomy (PPG I) has been widely accepted and used (cited 475 times on Google Scholar as of writing).
However, there were clearly problematic (non-monophyletic) genera at the time PPG I was published, and many new taxonomic changes have been (and will continue to be) made since [e.g., @Almeida2017; @Shang2018; @Zhang2020].
It is anticipated that "PPG II" will be a online, open resource that can be updated as necessary (Schuettpelz, pers. comm.).
We envision an open, continuously updated global fern phylogeny that is directly integrated with PPG II so that taxonomic decisions reflecting phylogeny can be made based on community consensus using the most recently available data.

Here, we leverage our specialized taxonomic knowledge to design a custom, fully reproducible, mostly automated pipeline to generate a maximally sampled global fern tree of life (FTOL).
We plan to run the pipeline on a semi-annual basis and make the results freely available online through a web portal and R package.
This will enable anybody interested in the biology of ferns to have access to the most recent hypothesis of fern phylogeny.
We anticipate FTOL will have several impacts for the field of fern systematics and evolution: 1) It will always provide the most up-to-date snapshot of our collective understanding of fern phylogeny; 2) It will allow for continuous assessment of taxonomy, and indicate what parts of the tree are in need of taxonomic revision; 3) It will be an important source of data for any phylogenetic comparative study of ferns.

# Materials and Methods

## Locus sampling

The plastid genome has been the most widely sequenced genomic compartment in ferns by far, so we used plastid loci to build our tree.
Our sampling includes two major categories of sequence data: seven loci (*atpA*, *atpB*, *matK*, *rbcL*, *rps4*, *trnL*--*trnF*, and *rps4*--*trnS*) that have been frequently used for molecular systematics studies in ferns and are typically obtained by PCR and Sanger sequencing ("Sanger loci") and a much larger set of single-copy, coding genes typically obtained from next-generation sequencing of the plastome ("plastome loci").
*trnL*--*trnF* includes the *trnL* intron, the 3' *trnL* exon, and the spacer between *trnL* and *trnF*.
*rps4*--*trnS* includes the spacer between *rps4* and *trnS*.
The set of plastome loci is based on the list of 83 protein-coding genes of @Wei2017a, which was filtered to only genes that show no evidence of duplication (77 genes), then combined with the two spacers (79 loci total).
The plastome loci include all seven Sanger loci.

## Sequence extraction

All sequences were downloaded from GenBank.
GenBank queries were formatted including terms "Polypodiopsida[ORGN]" to target ferns and "[PDAT]" to specify sequence date range (e.g., 1980/01/01[PDAT]:2021/12/12[PDAT]) so that a given query always results in the same set of sequences regardless of when the query is made.
There is no formal definition of genomic vs. Sanger sequences in GenBank, so we used an empirical sequence length cutoff to distinguish between Sanger (=< 7000 bp) or plastome sequences (> 7000 bp) with the "[SLEN]" term.
Furthermore, there is a lack of consensus on sequence annotation in GenBank; the same locus may be annotated using different names, or not annotated at all.

To avoid missing sequences due to differences in annotation format, we used the "Reference_Blast_Extract.py" script of  superCRUNCH [@Portik2020] to extract target loci from GenBank FASTA files.
Briefly, this involves querying candidate FASTA files downloaded from GenBank with BLAST [@Altschul1997] against a reference database of full length, representative sequences [i.e., a "baited search"\; @Smith2009b; @Smith2019].
Those portions of the query that have a significant match in the reference are extracted and written to a new filtered FASTA file.

We constructed the superCRUNCH reference databases by first downloading fern sequences from GenBank, then extracting target gene sequences with a custom R script that parses the GenBank flatfile; this only works for properly annotated accessions.
We then filtered the sequences to a single representative longest sequence per genus.
Next, we aligned the filtered sequences with MAFFT [@Katoh2002] and removed poorly aligned regions with trimAl [@Capella-Gutierrez2009].
To maximize the size of the reference database for Sanger loci, we then ran "Reference_Blast_Extract.py" using these sequences as references, followed by filtering to the longest sequence per genus and alignment and cleaning as before; this retrieved additional sequences that lacked annotations in the first round.
The cleaned alignments were then used as references for superCRUNCH to obtain a maximally sampled set of fern sequences from sequences downloaded from GenBank.

## Taxonomic name resolution

One goal of this project is generate a phylogeny that is consistent with the community-driven Pteridophyte Phylogeny Group I taxonomic system [@PteridophytePhylogenyGroupI2016\; hereafter "PPGI"].
Species names in GenBank do not necessarily conform to PPGI.
Therefore, we standardized all species names in the GenBank sequences against a reference taxonomy.
We designed the reference taxonomy for this project based on the World Ferns database [@Hassler2022], which conforms to PPGI and is available in Darwin Core format [@DarwinCore2009] via the Catalog of Life [@COL2366].
We used the taxastand R package [@Nitta2021] to resolve species names, which can account for differences in formatting of taxonomic authors (e.g., parenthetical basionym author present or absent) as well as perform fuzzy matching, which is needed to account for spelling errors or variations in author names.
We manually inspected any fuzzily matched or non-matching names.
This revealed some species names in GenBank that were completely lacking in the World Ferns database, spelling errors, as well as some names in the database that needed to be treated differently (e.g., changes in synonymy).
Therefore, we updated and edited the World Ferns database using the dwctaxon R package, which is designed to work with taxonomic data in the Darwin Core format.
We refer to the resulting taxonomic database as "pteridocat", and have made it freely available online (https://github.com/joelnitta/pteridocat) so that other researchers may standardize taxonomic names in their data to match those of FTOL.

## Automated removal of contaminations

Another potential issue with GenBank sequences is the presence of misidentified sequences and contaminations (hereafter "contaminations").
We removed putative contaminations from Sanger sequences as follows.
First, we constructed a BLAST library including all extracted Sanger sequences.
Next, we conducted one BLAST search for each Sanger sequence against the library (all-by-all BLAST).
We compared the families (following PPGI) of the matching sequences to each query; in the case the top three best matches belonged to a different family than the query, that query was considered a contamination and excluded from further analysis.
For Cyatheales and Saccolomatineae, which include several closely related small families, we used the order and sub-order level, respectively, instead of family.
While this method cannot account for contaminations at finer taxonomic levels, it is an efficient method to remove clearly incorrect sequences.

## Sequence selection and concatenation

A typical step in multilocus workflows is to concatenate loci across samples.
For phylogenetic studies aiming to produce one tip per species, this is often done by concatenating the longest sequence per locus within each species.
This is potentially problematic because accessions on GenBank may be misidentified and/or species may not be monophyletic.
Therefore, we concatenated accessions and selected one final set of concatenated loci per species as follows (here, we refer to "accession" to mean a single locus within a GenBank accession; GenBank accessions sometimes have multiple loci per accession, but we already split those out using superCRUNCH as described previously).
First, we constructed gene trees with FastTree [@Price2009; @Price2010] on default settings, and classified species as monophyletic if they were monophyletic in all gene trees including multiple accessions of that species.
We then concatenated loci for monophyletic species by selecting the longest accession per locus within each species.
Next, for the remaining species, we concatenated loci across accessions if all accessions originated from the same voucher specimen; this could not be done in all cases as voucher information is not always annotated on GenBank.
Next, for the remaining species, we concatenated loci across accessions if all accessions for a given species originated from only one publication, on the assumption that the authors of the study intended to do so.
For the remaining species that could not be concatenated using any of the above methods, we selected one longest accession per locus and discarded the others.

We then selected the final set of concatenated loci for each species. 
We sought to maximize representation of *rbcL* as this is historically the most sequenced locus for ferns, and we expect that maximal sampling of one locus should improve phylogenetic analysis.
Our selection is based on the following criteria, in order of priority:

1. The sample (set of accessions joined by the methods described above) with the longest concatenated sequence length; sample includes *rbcL* and additional loci.
2. The sample with the longest *rbcL* sequence; sample includes only *rbcL* and no additional loci.
3. The sample with the longest concatenated sequence length; sample does not include *rbcL*.

The above steps were not needed for plastome data, as each plastome sequence on GenBank originates from a single voucher specimen.
For these, we selected one specimen per species with the longest combined sequence length across all plastome loci.

## Sequence alignment

For non-spacer regions, we aligned each locus separately with MAFFT.

Spacer regions are difficult to align across higher taxonomic levels within ferns (e.g., family or higher) as they include frequent indels; however, spacer regions are very useful for phylogenetic analysis at finer levels (e.g., within genera) where slower-evolving genes like *rbcL* may not provide enough resolution.
Therefore, we first aligned spacer regions for each family with MAFFT, then merged the subalignments using the MAFFT "--merge" option.
This retains the indels within each subalignment while aligning across subalignments.
For some small families with fewer than three species each that could be used for subalignments, these were added as "singletons" during the MAFFT "--merge".
Anemiaceae showed an extremely high number of indels compared to other families and could not be reliably aligned across families, so we excluded it from the spacer region data.
We also excluded outgroups from the spacer region data as they cannot be reliably aligned to ferns.

Next, we removed poorly aligned regions from the alignments using trimAl.
We removed regions with >1% or >5% of sequences having gaps from coding regions or spacers, respectively.

## Phylogenetic analysis

Phylogenetic analysis was conducted in two steps.
First, we generated a backbone phylogeny using maximum likelihood (ML) analysis of the plastome loci in IQ TREE [@Nguyen2015] with 1000 ultrafast rapid bootstrap replicates [@Minh2013].
We applied the 'GTR+I+G' to a concatenated dataset with no partitioning.

Next, we used the backbone phylogeny as a constraint tree to conduct phylogenetic analysis of the Sanger loci in IQ TREE.
Settings for the ML analysis of the Sanger loci were otherwise identical to the backbone phylogenetic analysis.

## Reproducibility

The workflow is managed in R [@RCoreTeam2020a] with the targets package [@Landau2021].
Code used to generate FTOL is available at https://github.com/joelnitta/ftol.
A docker image to run the code is available at https://hub.docker.com/r/joelnitta/ftol.

# References
<!-- Insert <div id="refs"></div> to place the refs here instead of at the end -->
<div id="refs"></div>
